src/back/spirv/parser.rs

```diff
@@ -0,0 +1,1545 @@
+                }
+            }
+
+            _ => unimplemented!("{:?}", entry_point.exec_model),
+        }
+
+        instruction
+    }
+
+    fn get_type_id(
+        &mut self,
+        arena: &crate::Arena<crate::Type>,
+        handle: &crate::Handle<crate::Type>,
+    ) -> Word {
+        match self.lookup_type.lookup_id(&handle) {
+            Some(word) => word,
+            None => {
+                let (instruction, id) = self.instruction_type_declaration(arena, *handle);
+
+                for words in instruction.to_words() {
+                    self.module.logical_layout.type_declarations.push(words);
+                }
+                id
+            }
+        }
+    }
+
+    fn get_constant_id(
+        &mut self,
+        handle: &crate::Handle<crate::Constant>,
+        ir_module: &crate::Module,
+    ) -> Word {
+        match self.lookup_constant.lookup_id(&handle) {
+            Some(word) => word,
+            None => {
+                let (instruction, id) = self.instruction_constant_type(*handle, ir_module);
+
+                for words in instruction.to_words() {
+                    self.module.logical_layout.constants.push(words);
+                }
+                id
+            }
+        }
+    }
+
+    fn get_global_variable_id(
+        &mut self,
+        arena: &crate::Arena<crate::Type>,
+        global_arena: &crate::Arena<crate::GlobalVariable>,
+        handle: &crate::Handle<crate::GlobalVariable>,
+    ) -> Word {
+        match self.lookup_global_variable.lookup_id(&handle) {
+            Some(word) => word,
+            None => {
+                let global_variable = &global_arena[*handle];
+                let (instruction, id) =
+                    self.instruction_global_variable(arena, global_variable, handle);
+
+                for words in instruction.to_words() {
+                    self.module.logical_layout.global_variables.push(words);
+                }
+                id
+            }
+        }
+    }
+
+    fn get_function_type(
+        &mut self,
+        ty: Option<crate::Handle<crate::Type>>,
+        arena: &crate::Arena<crate::Type>,
+    ) -> Word {
+        match ty {
+            Some(handle) => self.get_type_id(arena, &handle),
+            None => match self.void_type {
+                Some(id) => id,
+                None => {
+                    let id = self.generate_id();
+
+                    let mut instruction = Instruction::new(Op::TypeVoid);
+                    instruction.set_result(id);
+
+                    self.void_type = Some(id);
+                    for word in instruction.to_words().iter() {
+                        self.module.logical_layout.type_declarations.push(*word);
+                    }
+                    id
+                }
+            },
+        }
+    }
+
+    fn find_scalar_handle(
+        &self,
+        arena: &crate::Arena<crate::Type>,
+        kind: &crate::ScalarKind,
```

Comment:
we should just pass things by value

Reply:
Could you explain maybe why? Because as far as I understood by reading through the Rust documentation, using references does not transfer ownership, and we wouldn't want to have ownership of these values in this function, or do we? Maybe I am not completely understanding this concept.

Reply:
Down to the hardware, the fastest way to call this function is via registers. Register needs to essentially store a copy of the data. If you are using a reference, you are telling it to work with a pointer, hoping that it will optimize this down to a register. This hope doesn't always materialize, and it's simpler to just tell it that you don't care about the original, aliasing, etc, you only need the value.

Reply:
Okay, that makes sense. You said somewhere in this PR that you only want to use references whenever the value will be bigger than 32 bits. So that would mean those optimization are greater than using references for values smaller than 32 bits?

Could I use this principle as guideline, so I won't put references everywhere? 😅

Reply:
A general principle is: only put stuff behind a ref if its size is greater than the size of a pointer (i.e. 64 bit).

Sometimes, we make exception from this. For example, all descriptors in wgpu-rs are passed by reference, regardless of their actual byte size.

Reply:
Thanks for all that information, that helps a lot!

Reply:
I've tried to remove the unnecessary references, but stumbled upon many move errors, especially for the `arena` in [`instruction_type_declaration`](https://github.com/Napokue/naga/blob/spirv-backend/src/back/spv/parser.rs#L351). The value is moved here, because I would give ownership to the `find_scalar_handle` function.

Reply:
Keep the references for stuff that is non-movable and isn't meant to be moved
My recommendation was simply not referencing trivial types like `ScalarKind`, since they are smaller than their references :)